File: src/route.py
from __future__ import annotations

import math
import random
from datetime import datetime, timedelta, timezone

from src.models import Coordinate

EARTH_RADIUS_M = 6_371_000


def haversine_distance_meters(start: Coordinate, end: Coordinate) -> float:
    lat1 = math.radians(start.lat)
    lat2 = math.radians(end.lat)
    dlat = math.radians(end.lat - start.lat)
    dlon = math.radians(end.lon - start.lon)

    a = math.sin(dlat / 2) ** 2 + math.cos(lat1) * math.cos(lat2) * math.sin(dlon / 2) ** 2
    return 2 * EARTH_RADIUS_M * math.atan2(math.sqrt(a), math.sqrt(1 - a))


def interpolate_points(start: Coordinate, end: Coordinate, step_meters: float) -> list[Coordinate]:
    if step_meters <= 0:
        raise ValueError("step_meters must be greater than zero")

    distance = haversine_distance_meters(start, end)
    if distance == 0:
        return [start]

    segments = max(1, math.ceil(distance / step_meters))
    points: list[Coordinate] = []
    for index in range(segments + 1):
        ratio = index / segments
        points.append(
            Coordinate(
                lat=start.lat + (end.lat - start.lat) * ratio,
                lon=start.lon + (end.lon - start.lon) * ratio,
                elevation=start.elevation + (end.elevation - start.elevation) * ratio,
            )
        )
    return points


def build_timed_route(
    points: list[Coordinate],
    speed_kmh: float,
    start_time: datetime | None = None,
    tick_seconds: int = 1,
) -> list[tuple[Coordinate, datetime]]:
    if not points:
        raise ValueError("at least one coordinate is required")
    if speed_kmh <= 0:
        raise ValueError("speed_kmh must be greater than zero")
    if tick_seconds <= 0:
        raise ValueError("tick_seconds must be greater than zero")

    current_time = start_time or datetime.now(timezone.utc)
    if current_time.tzinfo is None:
        current_time = current_time.replace(tzinfo=timezone.utc)

    speed_mps = speed_kmh * 1000 / 3600
    step_meters = speed_mps * tick_seconds
    timed: list[tuple[Coordinate, datetime]] = [(points[0], current_time)]

    for segment_start, segment_end in zip(points, points[1:]):
        segment_points = interpolate_points(segment_start, segment_end, step_meters)
        for point in segment_points[1:]:
            current_time += timedelta(seconds=tick_seconds)
            timed.append((point, current_time))

    return timed


def jitter_coordinate(point: Coordinate, radius_meters: float, rng: random.Random | None = None) -> Coordinate:
    if radius_meters <= 0:
        return point

    generator = rng or random.Random()
    distance = generator.uniform(0, radius_meters)
    bearing = generator.uniform(0, 2 * math.pi)
    north_meters = math.cos(bearing) * distance
    east_meters = math.sin(bearing) * distance

    lat_offset = north_meters / 111_320
    lon_scale = max(0.01, math.cos(math.radians(point.lat)))
    lon_offset = east_meters / (111_320 * lon_scale)
    return Coordinate(lat=point.lat + lat_offset, lon=point.lon + lon_offset, elevation=point.elevation)


def apply_route_jitter(
    timed_points: list[tuple[Coordinate, datetime]],
    radius_meters: float,
    seed: int | None = None,
) -> list[tuple[Coordinate, datetime]]:
    if radius_meters <= 0 or len(timed_points) < 3:
        return timed_points

    rng = random.Random(seed)
    jittered: list[tuple[Coordinate, datetime]] = [timed_points[0]]
    for point, timestamp in timed_points[1:-1]:
        jittered.append((jitter_coordinate(point, radius_meters, rng), timestamp))
    jittered.append(timed_points[-1])
    return jittered

File: internal/core/route.go
package core

import (
	"errors"
	"math"
	"math/rand"
	"time"
)

const EarthRadiusMeters = 6371000.0

func HaversineDistanceMeters(start, end Coordinate) float64 {
	lat1 := degreesToRadians(start.Lat)
	lat2 := degreesToRadians(end.Lat)
	dLat := degreesToRadians(end.Lat - start.Lat)
	dLon := degreesToRadians(end.Lon - start.Lon)

	a := math.Sin(dLat/2)*math.Sin(dLat/2) +
		math.Cos(lat1)*math.Cos(lat2)*math.Sin(dLon/2)*math.Sin(dLon/2)
	return 2 * EarthRadiusMeters * math.Atan2(math.Sqrt(a), math.Sqrt(1-a))
}

func InterpolatePoints(start, end Coordinate, stepMeters float64) ([]Coordinate, error) {
	if stepMeters <= 0 {
		return nil, errors.New("stepMeters must be greater than zero")
	}

	distance := HaversineDistanceMeters(start, end)
	if distance == 0 {
		return []Coordinate{start}, nil
	}

	segments := int(math.Ceil(distance / stepMeters))
	if segments < 1 {
		segments = 1
	}

	points := make([]Coordinate, 0, segments+1)
	for i := 0; i <= segments; i++ {
		ratio := float64(i) / float64(segments)
		points = append(points, Coordinate{
			Lat:       start.Lat + (end.Lat-start.Lat)*ratio,
			Lon:       start.Lon + (end.Lon-start.Lon)*ratio,
			Elevation: start.Elevation + (end.Elevation-start.Elevation)*ratio,
		})
	}

	return points, nil
}

func BuildTimedRoute(points []Coordinate, speedKmh float64, start time.Time, tick time.Duration) ([]TimedPoint, error) {
	if len(points) == 0 {
		return nil, errors.New("at least one coordinate is required")
	}
	if speedKmh <= 0 {
		return nil, errors.New("speedKmh must be greater than zero")
	}
	if tick <= 0 {
		return nil, errors.New("tick must be greater than zero")
	}
	if start.IsZero() {
		start = time.Now().UTC()
	}
	start = start.UTC()

	speedMps := speedKmh * 1000 / 3600
	stepMeters := speedMps * tick.Seconds()
	current := start
	timed := []TimedPoint{{Point: points[0], Time: current.Unix()}}

	for i := 0; i < len(points)-1; i++ {
		segmentPoints, err := InterpolatePoints(points[i], points[i+1], stepMeters)
		if err != nil {
			return nil, err
		}
		for _, point := range segmentPoints[1:] {
			current = current.Add(tick)
			timed = append(timed, TimedPoint{Point: point, Time: current.Unix()})
		}
	}

	return timed, nil
}

func JitterCoordinate(point Coordinate, radiusMeters float64, rng *rand.Rand) Coordinate {
	if radiusMeters <= 0 {
		return point
	}
	if rng == nil {
		rng = rand.New(rand.NewSource(time.Now().UnixNano()))
	}

	distance := rng.Float64() * radiusMeters
	bearing := rng.Float64() * 2 * math.Pi
	northMeters := math.Cos(bearing) * distance
	eastMeters := math.Sin(bearing) * distance

	latOffset := northMeters / 111320
	lonScale := math.Max(0.01, math.Cos(degreesToRadians(point.Lat)))
	lonOffset := eastMeters / (111320 * lonScale)

	return Coordinate{
		Lat:       point.Lat + latOffset,
		Lon:       point.Lon + lonOffset,
		Elevation: point.Elevation,
	}
}

func ApplyRouteJitter(points []TimedPoint, radiusMeters float64, seed int64) []TimedPoint {
	if radiusMeters <= 0 || len(points) < 3 {
		return points
	}

	var rng *rand.Rand
	if seed != 0 {
		rng = rand.New(rand.NewSource(seed))
	}

	jittered := make([]TimedPoint, len(points))
	copy(jittered, points)
	for i := 1; i < len(points)-1; i++ {
		jittered[i].Point = JitterCoordinate(points[i].Point, radiusMeters, rng)
	}

	return jittered
}

func TimedPointsCoordinates(points []TimedPoint) []Coordinate {
	coords := make([]Coordinate, len(points))
	for i, point := range points {
		coords[i] = point.Point
	}
	return coords
}

func OffsetCoordinate(point Coordinate, eastMeters, northMeters float64) Coordinate {
	latOffset := northMeters / 111320
	lonScale := math.Max(0.01, math.Cos(degreesToRadians(point.Lat)))
	lonOffset := eastMeters / (111320 * lonScale)
	return Coordinate{Lat: point.Lat + latOffset, Lon: point.Lon + lonOffset, Elevation: point.Elevation}
}

func degreesToRadians(value float64) float64 {
	return value * math.Pi / 180
}
